### DISCUSSION TITLE
Allow `flux build/diff` to optionally expand the Helm release

### DISCUSSION THREAD

**Discussion by rombert** [fluxcd/flux2#2434] [Proposals] [answered]
Hi,

I think the `flux build` and `flux diff` commands are very useful when allowing to locally preview will be applied in the cluster, whether it's for CI or for local development.

These commands would be even more useful if they would expand the HelmRelease objects and render the resources based on the templates. This way we would get a more accurate representation of the resources that will be applied.

One situation where this would've been useful a bit of time ago was detecting Helm charts were were using a deprecated apiVersion for Ingress resources. And, of course, this opens up the possibility to run all kinds of static analysers, which some may prefer to running checks in-cluster.

**Comment by kingdonb** [comment, marked as answer]
It would be good to add support for this, as users who are validating a PR that updates a HelmRelease currently have a gap if they do not have confidence through other systems of support that their Helm Charts are definitely valid. This can also be an important verification of `postRender` Kustomizations in a HelmRelease that cannot so easily be verified outside of Flux.

There are at least three separate parts to this idea:

1. support HelmRelease in flux build/diff
2. Let `flux build/diff` traverse the tree so `flux build` on a Kustomization also builds the Kustomization that it applies downstream, and any HelmRelease downstream, (and any others? These are the only Flux Appliers, so that's probably where I'd draw the line...)
3. A configurable depth or level option, something like `--level` or `-L` from the unix `tree` command, to set the depth of traversal.

Say you have a tree that is 4 levels deep, but you only want the top three levels expanded because dev-teams are responsible for their own resources. Maybe a "level" parameter is not the best/only way to provide a scope or depth limiting behavior.

For unix tree, the default behavior assumes you want an unlimited depth; I think this would be a reasonable default for us too, if we are expanding the tree in build, so that you can validate everything at once. Maybe there are other reasons or approaches to set a limit.

Sounds like a very useful idea!

**Reply by stefanprodan** [reply]
I think this is a duplicate of https://github.com/fluxcd/flux2/discussions/820
